### PR TITLE
AK+LibC: Fix performance issues with large time_t values

### PR DESCRIPTION
**Pull request by BertalanD**
Fixes 12729

### CODE REVIEW

**Comment by nico** [comment]
Needs `#` in front of the number on the commit message, else merging this won't auto close the issue I think.

File: AK/Time.h [hunk 1]
```diff
@@ -56,24 +56,40 @@ constexpr int day_of_year(int year, unsigned month, int day)
...
 
-constexpr int years_to_days_since_epoch(int year)
+constexpr i64 years_to_days_since_epoch(i64 year)
 {
-    int days = 0;
-    for (int current_year = 1970; current_year < year; ++current_year)
-        days += days_in_year(current_year);
-    for (int current_year = year; current_year < 1970; ++current_year)
-        days -= days_in_year(current_year);
-    return days;
```
Comment (nico) [inline AK/Time.h:73]: The LHS code here is much shorter and easier to understand. Can't we keep that?

Comment (BertalanD) [reply AK/Time.h:73]: The issue here is that the current algorithm is linear-time, and ends up producing **huge** run times when called with large values.

As per the (unsuccessfully) linked issue, some calculations performed by `configure` scripts take more than a minute.

File: Userland/Libraries/LibC/time.cpp [hunk 2]
```diff
@@ -89,15 +89,37 @@ char* ctime_r(const time_t* t, char* buf)
     return asctime_r(localtime_r(t, &tm_buf), buf);
 }
 
-static const int __seconds_per_day = 60 * 60 * 24;
+static constexpr i64 __seconds_per_day = 60 * 60 * 24;
+static constexpr i64 __seconds_per_4_years = (3 * 365 + 366) * __seconds_per_day;
+static constexpr i64 __seconds_per_100_years = (76 * 365 + 24 * 366) * __seconds_per_day;
+static constexpr i64 __seconds_per_400_years = (303 * 365 + 97 * 366) * __seconds_per_day;
```
Comment (BertalanD) [inline Userland/Libraries/LibC/time.cpp:95]: I'm probably better off putting these constants in `AK/Time.h` to reduce duplication between that and `time.cpp`.

**Comment by bgianfo** [comment]
Can you try this test from perl? I was able to confirm that this causes our `gmtime` to spin for what appears to be forever (maybe just a very long time). 

https://raw.githubusercontent.com/Perl/perl5/3e25f6d2dce9289bf3894d673f992195fb530d8e/Porting/timecheck2.c

**Comment by bgianfo** [comment]
> Can you try this test from perl? It runs for ~10 minutes on serenity right now, and runs for in ~0.5ms on my Linux host.

I tried it my self with your changes, and it seems to run in ~95 ms now :partying_face:.

File: AK/Time.h [hunk 3]
```diff
@@ -56,24 +56,40 @@ constexpr int day_of_year(int year, unsigned month, int day)
...
+int days_in_month(i64 year, unsigned month);
 
-constexpr int days_in_year(int year)
+constexpr int days_in_year(i64 year)
 {
     return 365 + (is_leap_year(year) ? 1 : 0);
 }
 
-constexpr int years_to_days_since_epoch(int year)
+constexpr i64 years_to_days_since_epoch(i64 year)
```
Comment (trflynn89) [inline AK/Time.h:66]: This might have a bug for the year 2100. Before this change, `years_to_days_since_epoch` returned 47482. Now it returns 47481.

Comment (trflynn89) [reply AK/Time.h:66]: After that is fixed, if you want a free commit, feel free to revert 032664332be6e1d42567165e9ffb623faf2b9a10 :)

File: Userland/Libraries/LibC/time.cpp (hunk 2 above)
Comment (AtkinsSJ) [reply Userland/Libraries/LibC/time.cpp:95]: We have a `AK/DateConstants.h` file too now, for what it's worth.

File: AK/Time.h (hunk 1 above)
Comment (ldm5180) [reply AK/Time.h:73]: In fact, you might not even be able to run the LHS in a constexpr context given a large enough value without changing the compiler settings.

Comment (greg904) [reply AK/Time.h:73]: What about something like that:
```cpp 
i64 leap_years = year / 400 + year / 4 - year / 1000;
i64 non_leap_years = year - leap_years;
i64 days_since_bce = non_leap_years * 365 + leap_years * 366;
i64 days_from_bce_to_epoch = 719527;
return days_since_bce - days_from_bce_to_epoch;
```

**Comment by BertalanD** [comment]
As Tim pointed out, this code is actually incorrect.

Instead of me trying to re-invent the wheel, a more practical approach would be to implement the algorithms described here: https://howardhinnant.github.io/date_algorithms.html, specifically `days_from_civil` and `civil_from_days`.

**Pull request closed without merging**